src/main/client/admin.c

```diff
@@ -31,124 +31,165 @@
- *******************************************************************************************************
- */
-PyObject *AerospikeClient_Admin_Create_User(AerospikeClient *self,
-                                            PyObject *args, PyObject *kwds)
+#define STR_CONVERSION_ERROR_MSG "%s should be a string"
+
+// If we are creating a PKI user, py_password can be NULL
+// If not creating a PKI user, py_password must be non-NULL
+// In both cases, the other pyobjects must be non-NULL
+static PyObject *admin_create_user_helper(AerospikeClient *self,
+                                          PyObject *py_user,
+                                          PyObject *py_password,
+                                          PyObject *py_roles,
+                                          PyObject *py_policy)
 {
-    // Initialize error
     as_error err;
     as_error_init(&err);
 
-    // Python Function Arguments
-    PyObject *py_policy = NULL;
-    PyObject *py_user = NULL;
-    PyObject *py_password = NULL;
-    PyObject *py_roles = NULL;
-
-    as_policy_admin admin_policy;
-    as_policy_admin *admin_policy_p = NULL;
-
-    // Python Function Keyword Arguments
-    static char *kwlist[] = {"user", "password", "roles", "policy", NULL};
-
-    // Python Function Argument Parsing
-    if (PyArg_ParseTupleAndKeywords(args, kwds, "OOO|O:admin_create_user",
-                                    kwlist, &py_user, &py_password, &py_roles,
-                                    &py_policy) == false) {
-        return NULL;
-    }
-
-    // Aerospike Operation Arguments
-    int roles_size = 0;
-    char **roles = NULL;
-    const char *user, *password;
-
     if (!self || !self->as) {
         as_error_update(&err, AEROSPIKE_ERR_PARAM, "Invalid aerospike object");
-        goto CLEANUP;
+        goto RAISE_EXCEPTION;
     }
 
     if (!self->is_conn_16) {
         as_error_update(&err, AEROSPIKE_ERR_CLUSTER,
                         "No connection to aerospike cluster");
-        goto CLEANUP;
+        goto RAISE_EXCEPTION;
     }
 
     // Convert python object to an array of roles
-    if (PyList_Check(py_roles)) {
-        roles_size = PyList_Size(py_roles);
-        roles = alloca(sizeof(char *) * roles_size);
-        for (int i = 0; i < roles_size; i++) {
-            roles[i] = cf_malloc(sizeof(char) * AS_ROLE_SIZE);
-            memset(roles[i], 0, sizeof(char) * AS_ROLE_SIZE);
-        }
-    }
-    else {
+    if (!PyList_Check(py_roles)) {
         as_error_update(&err, AEROSPIKE_ERR_PARAM, "Roles should be a list");
-        goto CLEANUP;
+        goto RAISE_EXCEPTION;
+    }
+    int roles_size = PyList_Size(py_roles);
+    if (PyErr_Occurred()) {
+        goto RAISE_EXCEPTION;
+    }
+    char **roles = alloca(sizeof(char *) * roles_size);
+    for (int i = 0; i < roles_size; i++) {
+        roles[i] = cf_malloc(sizeof(char) * AS_ROLE_SIZE);
+        memset(roles[i], 0, sizeof(char) * AS_ROLE_SIZE);
```

Comment:
Why allocate the container on the stack and the role items on the heap? Can both go on the stack?

Reply:
I just reused the code from before where it was allocating a a list of strings on the stack, and each string was being heap allocated. I think the original maintainer used "alloca" because it was faster than to heap allocate memory, and my understanding is using "alloca" to allocate too much memory can lead to a stack overflow.

I was going to refactor the whole admin.c since it's using copy and pasted code to allocate a list of strings. This was out of the scope of this PR, so I left it out for now